### ISSUE TITLE
[Stacks 2.1] Regressions in Clarity 1 type checking

### ISSUE THREAD

**Issue by lgalabru** [stacks-network/stacks-blockchain#3426]
In the process of releasing v1.2.0 of Clarinet - pulling the very latest next vm of this repo - a few regressions in the way Clarity v1 is supposed to handle type checking were revealed.

- Case 1

```clarity
(define-public (caller) 
    (begin 
        (asserts! (is-eq  .callee contract-caller) (err u1))
        (ok u1)))
```
fails with 
```
error: expecting expression of type '(principal ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.callee)', found 'principal'
```

- Case 2

```clarity
(define-public (caller) 
    (begin 
        (asserts! (is-eq  contract-caller .callee) (err u1))
        (ok u1)))
```
fails with 
```
error: expecting expression of type 'principal', found '(principal ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.callee)'
```

- Case 3

```clarity
(begin 
    (contract-call? .base-dao set-extensions
      (list
        {extension: .ccd001-direct-execute, enabled: true}
        {extension: .ccd002-treasury-mia, enabled: true}
        {extension: .ccd002-treasury-nyc, enabled: true}
    )
))

(define-public (set-extensions (extensionList (list 200 {extension: principal, enabled: bool})))
...
)
```

fails with
```
expecting expression of type '(list 200 (tuple (enabled bool) (extension principal)))', found '(list 3 (tuple (enabled bool) (extension principal)))'
```

This 3 issues were patch directly in https://github.com/stacks-network/stacks-blockchain/pull/3293 to unblock the Clarinet release
- Case 1 and 2 patched with https://github.com/stacks-network/stacks-blockchain/pull/3293/commits/e8ad18055de64db6fbe02914b95367c20bcb9212
- Case 3 patched with https://github.com/stacks-network/stacks-blockchain/pull/3293/commits/601ac7a0b4ed3de027382a33560d079aaa337b4d

These patches need to have their dedicated PR with unit testing, then 2 aforementioned commits will be reverted from that PR.

I think we want to spin up a Stacks 2.1 mainnet node, booting from network, that would help convince us that all the changes that were introduced in `next` are backward compatible - a stacks 2.1 node with a very distant epoch 2.1 transition should be able to produce the same chainstate than a stacks-node coming from mainline.

**Comment by jcnelson** [comment]
@obycode Can you take a look at this?

**Comment by obycode** [comment]
Yup, I'm on it. Discussed with Ludo yesterday.

**Comment by jcnelson** [comment]
I really think we should be fuzz-testing this.  Specifically, we should be running a fuzzer that compares the AST expressions produced from the v1 parser in `master` and the expressions produced by the one in `next`.  Unfortunately, this kind of workload seems to be outside what `cargo-fuzz` can do, because it will require comparing the behavior of two different `git` branches.

Related, we really should have fuzz targets for both the v1 and v2 parsers running 24/7.  This is something Cargo can easily do.

**Comment by obycode** [comment]
Agreed, fizz testing would be great to have here. Note that the problem seen here is not from the parser changes, but is from the type-checker changes from the trait work.

**Comment by jcnelson** [comment]
Right; I'm suggesting we examine the resulting AST expressions from the `ast::build_ast_with_rules()` function in both cases, which covers the new trait implementation and everything else.

**Comment by jcnelson** [comment]
We need to cherry-pick the fixes into `next`.

**Comment by jcnelson** [comment]
This is fixed by #3432
